teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md

```diff
@@ -0,0 +1,239 @@
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
+
+None
+
+### User Experience (optional)
+
+<!--
+Consideration about the user experience. Depending on the area of change,
+users may be task and pipeline editors, they may trigger task and pipeline
+runs or they may be responsible for monitoring the execution of runs,
+via CLI, dashboard or a monitoring system.
+
+Consider including folks that also work on CLI and dashboard.
+-->
+
+With the embedded taskSpec for the custom task, all the Tekton clients
+can create a pipeline or pipelineRun using a single API call to the Kubernetes.
+Any downstream systems that employ tektoncd e.g. Kubeflow pipelines, will not be
+ managing all the custom task CRs and their versioning.
+
+It is natural for a user to follow ways such as a 
+[Kubernetes Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), ReplicaSet, StatefulSet 
+and also Tektoncd/Pipeline taskSpec to have a Pipeline with custom tasks embedded.
+
+### Performance (optional)
+
+<!--
+Consideration about performance.
+What impact does this change have on the start-up time and execution time
+of task and pipeline runs? What impact does it have on the resource footprint
+of Tekton controllers as well as task and pipeline runs?
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+
+If it's helpful to include workflow diagrams or any other related images,
+add them under "/teps/images/". It's upto the TEP author to choose the name
+of the file, but general guidance is to include at least TEP number in the
+file name, for example, "/teps/images/NNNN-workflow.jpg".
+-->
+
+## Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy.  Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+-->
+
+We can reuse the custom task e2e tests with the current test controller
+to verify whether the controller can handle the custom task taskSpec.
```

Comment:
Would you then add support for embedded spec to the test controller?

Reply:
We will update the custom task test over here. Looks like there no test controller running in the e2e test?
https://github.com/tektoncd/pipeline/blob/main/test/custom_task_test.go